### PR TITLE
[SPARK-40137][SQL] Combines limits after projection

### PR DESCRIPTION
**Pull request by ConeyLiu**
### What changes were proposed in this pull request?

Combine the limits after projection.

### Why are the changes needed?

`Dataset.show` will add extra `Limit` and `Projection` on top of the given logical plan. If the `Dataset` is already a limit job, this will introduce an extra shuffle phase. So we should combine the limit after projection.

For example:
```scala
spark.sql("select * from spark.store_sales limit 10").show()
```

Before:
```
== Physical Plan ==
AdaptiveSparkPlan (12)
+- == Final Plan ==
   * Project (7)
   +- * GlobalLimit (6)
      +- ShuffleQueryStage (5), Statistics(sizeInBytes=185.6 KiB, rowCount=990)
         +- Exchange (4)
            +- * LocalLimit (3)
               +- * ColumnarToRow (2)
                  +- Scan parquet spark_catalog.spark.store_sales (1)
+- == Initial Plan ==
   Project (11)
   +- GlobalLimit (10)
      +- Exchange (9)
         +- LocalLimit (8)
            +- Scan parquet spark_catalog.spark.store_sales (1)
```
After:
```
== Physical Plan ==
CollectLimit (4)
+- * Project (3)
   +- * ColumnarToRow (2)
      +- Scan parquet spark_catalog.spark.store_sales (1)
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT and local test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1994,6 +1994,13 @@ object EliminateLimits extends Rule[LogicalPlan] {
       LocalLimit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
     case Limit(le, Limit(ne, grandChild)) =>
       Limit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
+
+    case Project(projectList, Limit(l, child)) =>
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1998]: `projectList` may contain very heavy UDFs. This change may increase `Project` evaluation cost.

Comment (ConeyLiu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1998]: Thanks @wangyum for your time. What about:
```scala
case Limit(l, Project(projectList, child)) =>
  Project(projectList, Limit(l, child))
```

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1998]: +1.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1994,6 +1994,13 @@ object EliminateLimits extends Rule[LogicalPlan] {
       LocalLimit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
     case Limit(le, Limit(ne, grandChild)) =>
       Limit(Literal(Least(Seq(ne, le)).eval().asInstanceOf[Int]), grandChild)
+
+    case GlobalLimit(l, Project(projectList, child)) =>
+      Project(projectList, GlobalLimit(l, child))
+    case LocalLimit(l, Project(projectList, child)) =>
+      Project(projectList, LocalLimit(l, child))
+    case Limit(l, Project(projectList, child)) =>
+      Project(projectList, Limit(l, child))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2003]: Could you move these changes to `LimitPushDown`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -769,6 +783,13 @@ object LimitPushDown extends Rule[LogicalPlan] {
     // Push down local limit 1 if join type is LeftSemiOrAnti and join condition is empty.
     case j @ Join(_, right, LeftSemiOrAnti(_), None, _) if !right.maxRows.exists(_ <= 1) =>
       j.copy(right = maybePushLocalLimit(Literal(1, IntegerType), right))
+
+    case GlobalLimit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, GlobalLimit(l, child))
+    case LocalLimit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, LocalLimit(l, child))
+    case Limit(l, proj @ Project(projectList, child)) if canPushLimitThroughProject(proj) =>
+      Project(projectList, Limit(l, child))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:792]: How about?
```scala
    case GlobalLimit(l, p @ Project(projectList, child))
        if child.outputSet.subsetOf(p.references) =>
      Project(projectList, GlobalLimit(l, child))
    case LocalLimit(l, p @ Project(projectList, child))
       if child.outputSet.subsetOf(p.references) =>
      Project(projectList, LocalLimit(l, child))
    case Limit(l, p @ Project(projectList, child))
        if child.outputSet.subsetOf(p.references) =>
      Project(projectList, Limit(l, child))
```

Comment (ConeyLiu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thanks, @wangyum. This is a really much better solution. The failed python ut should be fixed by this update.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -769,6 +769,16 @@ object LimitPushDown extends Rule[LogicalPlan] {
...
+
+    case GlobalLimit(l, p @ Project(projectList, child))
+      if child.outputSet.subsetOf(p.references) =>
+      Project(projectList, GlobalLimit(l, child))
+    case LocalLimit(l, p @ Project(projectList, child))
+      if child.outputSet.subsetOf(p.references) =>
+      Project(projectList, LocalLimit(l, child))
+    case Limit(l, p @ Project(projectList, child))
+      if child.outputSet.subsetOf(p.references) =>
+      Project(projectList, Limit(l, child))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:781]: So we are trying to optimize this case?
```scala
spark.sql("select *,ss_sold_time_sk + 3  from (select *, ss_sold_time_sk + 2 from (select *, ss_sold_time_sk + 1 from store_sales limit 10) limit 9) limit 8").show()
```

Comment (ConeyLiu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:781]: Yes, like this.

**Pull request closed without merging**